src/plugins/p/memory-darwin.c

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2015 Steve Schnepp <steve.schnepp@pwkf.org> - All rights reserved.
+ *
+ * This copyrighted material is made available to anyone wishing to use,
+ * modify, copy, or redistribute it subject to the terms and conditions
+ * of the GNU General Public License v.2 or v.3.
+ */
+
+#include <mach/mach_host.h>
+#include <mach/vm_statistics.h>
+#include <stdio.h>
+#include <sys/sysctl.h>
+#include "common.h"
+#include "plugins.h"
+
+int memory(int argc, char **argv) {
+
+	if(argc > 1) {
+		if(!strcmp(argv[1], "config")) {
+			puts("graph_vlabel Bytes\n"
+				 "graph_args --base 1024 -l 0\n"
+				 "graph_title Memory usage\n"
+				 "graph_category system\n"
+				 "graph_info This graph shows what the machine uses memory for.\n"
+
+				 "wired.label Wired\n"
+				 "wired.draw AREA\n"
+				 "wired.info Nonswappable memory used by the kernel.\n"
+
+				 "active.label Active\n"
+				 "active.draw STACK\n"
+				 "active.info Recently-accessed memory.\n"
+
+				 "inactive.label Inactive\n"
+				 "inactive.draw STACK\n"
+				 "inactive.info Memory likely to be purged or swapped out.\n"
+
+				 "speculative.label Speculative\n"
+				 "speculative.draw STACK\n"
+				 "speculative.info File-backed/mmapped memory, speculatively read and cached.\n"
+
+				 "other.label Other\n"
+				 "other.draw STACK\n"
+
+				 "free.label Free\n"
+				 "free.draw STACK\n"
+				 "free.info Available for immediate use.");
+
+			return 0;
+		}
+		if(!strcmp(argv[1], "autoconf"))
+			return writeyes();
+	}
+
+	uint32_t pagesize;
+	size_t len = sizeof(pagesize);
+	if (sysctlbyname("vm.pagesize", &pagesize, &len, NULL, 0) < 0) {
+		return fail("sysctlbyname");
+	}
+
+	uint32_t totalPages;
+	len = sizeof(totalPages);
+	if (sysctlbyname("vm.pages", &totalPages, &len, NULL, 0) < 0) {
+		return fail("sysctlbyname");
+	}
+
+	struct vm_statistics64 stats;
+	host_t host = mach_host_self();
+	mach_msg_type_number_t count = HOST_VM_INFO64_COUNT;
+	if (host_statistics64(host, HOST_VM_INFO64, (host_info64_t)&stats, &count) != KERN_SUCCESS)  {
+		return fail("host_statistics64");
+	}
+
+	printf("wired.value %lu\n", stats.wire_count * pagesize);
+	printf("active.value %lu\n", stats.active_count * pagesize);
+	printf("inactive.value %lu\n", stats.inactive_count * pagesize);
+	printf("speculative.value %lu\n", stats.speculative_count * pagesize);
+
+	int64_t freePages = stats.free_count - stats.speculative_count;
+	printf("other.value %lu\n", MAX(0, (int64_t)totalPages - stats.wire_count - stats.active_count - stats.inactive_count - freePages) * pagesize);
```

Comment:
Do not hesitate to use temporary named variables.
They get optimized away and help understanding and debugging: 😉

Reply:
Oh, yes, definitely.  I think the problem here is that I don't understand how these numbers are supposed to add up, and I haven't been able to find complete documentation on them.  I found https://github.com/apple/darwin-xnu/blob/2ff845c2e033bd0ff64b5b6aa6063a1f8f65aa32/osfmk/mach/vm_statistics.h#L108-L111 which says:

> speculative pages are already accounted for in "free_count", so "speculative_count" is the number of "free" pages that are used to hold data that was read speculatively from disk but haven't actually been used by anyone so far.

…so I've defined `freePages = stats.free_count - stats.speculative_count`.

But often the values in `vm_statistics` don't sum to exactly `vm.pages` — sometimes the sum is less than `vm.pages` (represented by the `other.value` I added); sometimes the sum is more than `vm.pages` (thus the need for `MAX(0, …)` to keep it from wrapping).

Here's what the latest revision (7b2914a) looks like for an 8 GB system:

![memory-day](https://user-images.githubusercontent.com/190164/189695519-65c52409-eb93-4456-9c27-75e5c277571d.png)